File: week3_greedy_algorithms/3_car_fueling/car_fueling.py
from sys import stdin


def min_refills(distance, tank, stops):
    # write your code here
    # m_refills = round(distance/tank)
    refills = 0
    pre_stop = 0
    for i in range(len(stops)):
        
        re_tank = tank- (stops[i]-pre_stop)
        re_dis = distance - stops[i]
        pre_stop = stops[i]

        if tank >= re_dis:
            return refills
        else:
            if re_tank < stops[i+1] - stops[i]:
                if tank < stops[i+1] - stops[i]:
                    return -1 
                refills += 1
    return refills

    

if __name__ == '__main__':
    # d, m, _, *stops = map(int, stdin.read().split())
    d = 15
    m = 5
    stops = [2,8,9,12]
    print(min_refills(d, m, stops))

File: week3_greedy_algorithms/3_car_fueling/car_fueling.cpp
#include <iostream>
#include <vector>

using std::cin;
using std::cout;
using std::vector;
using std::max;

int compute_min_refills(int dist, int tank, vector<int> & stops) {
    // write your code here
    int refills = 0;
    int pre_stop = 0;
    int re_tank = 0;
    int re_dis = 0;

    for (int i; i<sizeof(stops); i++ )
    {
        re_tank = tank - (stops[i] - pre_stop);
        re_dis = dist -stops[i];

        pre_stop = stops[i];

        if (tank >= re_dis)
            return refills;
        else
        {
            if (re_tank < (stops[i+1] - stops[i]))
                {
                    if (tank < (stops[i+1] - stops[i]))
                        return -1;
                    refills += 1;
                }       
        }
    }
    return refills;
}


int main() {
    int d = 0;
    cin >> d;
    int m = 0;
    cin >> m;
    int n = 0;
    cin >> n;

    vector<int> stops(n);
    for (size_t i = 0; i < n; ++i)
        cin >> stops.at(i);

    cout << compute_min_refills(d, m, stops) << "\n";

    return 0;
}
